Mathlib/CategoryTheory/WithTerminal/Cone.lean

```diff
@@ -0,0 +1,225 @@
+`WithTerminal J ⥤ C`, that sends `star` to `X`. -/
+@[simps!]
+def liftFromOver : (J ⥤ Over X) ⥤ WithTerminal J ⥤ C := commaFromOver ⋙ equivComma.inverse
+
+/-- The extension of a functor to over categories behaves well with compositions. -/
+@[simps]
+def liftFromOverComp : liftFromOver.obj (K ⋙ Over.post F) ≅ liftFromOver.obj K ⋙ F where
+  hom.app | star | of a => 𝟙 _
+  inv.app | star | of a => 𝟙 _
+
+/-- A cone of a functor `K : J ⥤ Over X` consists of an object of `Over X`, together
+with morphisms. This same object is a cone of the extended functor
+`liftFromOver.obj K : WithTerminal J ⥤ C`. -/
+@[simps]
+private def coneLift : Cone K ⥤ Cone (liftFromOver.obj K) where
+  obj t := {
+    pt := t.pt.left
+    π.app
+    | of a => (t.π.app a).left
+    | star => t.pt.hom
+    π.naturality
+    | star, star, _
+    | of a, star, _ => by aesop
+    | of a, of b, f => by simp [← Comma.comp_left]
+  }
+  map {t₁ t₂} f := {
+    hom := f.hom.left
+    w
+    | star => by aesop_cat
+    | of a => by simp [← Comma.comp_left]
+  }
+
+/-- This is the inverse of the previous construction: a cone of an extended functor
+`liftFromOver.obj K : WithTerminal J ⥤ C` consists of an object of `C`, together
+with morphisms. This same object is a cone of the original functor `K : J ⥤ Over X`. -/
+@[simps]
+private def coneBack : Cone (liftFromOver.obj K) ⥤ Cone K where
+  obj t := {
+    pt := .mk (t.π.app star)
+    π.app a := {
+      left := t.π.app (of a)
+      right := 𝟙 _
+      w := by simpa using t.w (homFrom a)
+    }
+    π.naturality a b f := by ext; simpa using t.π.naturality (incl.map f)
+  }
+  map {t₁ t₂ f} := {
+    hom := Over.homMk f.hom
+  }
+
+/-- Given a functor `K : J ⥤ Over X` and its extension `liftFromOver K : WithTerminal J ⥤ C`,
+there is an obvious equivalence between cones of these two functors.
+A cone of `K` is an object of `Over X`, so it has the form `t ⟶ X`.
+Equivalently, a cone of `WithTerminal K` is an object `t : C`,
+and we can recover the structure morphism as `π.app X : t ⟶ X`. -/
+@[simps! functor_obj_pt functor_map_hom inverse_obj_pt_left inverse_obj_pt_right_as
+  inverse_obj_pt_hom inverse_obj_π_app_left inverse_map_hom_left unitIso_hom_app_hom_left
+  unitIso_inv_app_hom_left counitIso_hom_app_hom counitIso_inv_app_hom]
+def coneEquiv : Cone K ≌ Cone (liftFromOver.obj K) where
+  functor := coneLift
+  inverse := coneBack
+  unitIso := .refl _
+  counitIso := NatIso.ofComponents fun t ↦ Cones.ext <| .refl _
+
+@[simp]
+theorem coneEquiv_functor_obj_π_app_star : (coneEquiv.functor.obj t).π.app star = t.pt.hom :=
+  by aesop
+
+@[simp]
+theorem coneEquiv_functor_obj_π_app_of (Y : J): (coneEquiv.functor.obj t).π.app (of Y) =
+    (t.π.app Y).left := by aesop
+
+/-- A cone `t` of `K : J ⥤ Over X` is a limit if and only if the corresponding cone
+`coneLift t` of `liftFromOver.obj K : WithTerminal K ⥤ C` is a limit. -/
+@[simps!]
+def limitEquiv : IsLimit (coneLift.obj t) ≃ IsLimit t := IsLimit.ofConeEquiv coneEquiv
```

Comment:
This is a public definition which uses a private definition in its type, which isn't generally a good idea.